build/openwisp_base/Dockerfile

```diff
@@ -43,7 +43,8 @@ ARG OPENWISP_UTILS_SOURCE=https://github.com/openwisp/openwisp-utils/tarball/doc
 RUN pip install --prefix='/install' --upgrade --force-reinstall ${OPENWISP_UTILS_SOURCE}
 RUN pip install --prefix='/install' channels_redis service_identity django-redis psycopg2 \
                                     uwsgi sentry-sdk supervisor celery django-celery-beat \
-                                    django-cors-headers django-rest-auth jinja2
+                                    django-cors-headers django-rest-auth jinja2 \
+                                    django-pipeline~=2.0.0 django-compress-staticfiles~=1.0.1b
```

Comment:
Can you please add a reply here. (not in code right now)
About what all I need to test when I am upgrading it in the future? 😄

Reply:
> Can you please add a reply here. (not in code right now) About what all I need to test when I am upgrading it in the future? smile

Yes, I'll update you :+1: 
One more thing, While I'm installing `docker-openwisp` for development using `make develop`, I'm getting this error :point_down: 

```bash
ERROR: HTTP error 404 while getting https://github.com/openwisp/netjsonconfig/tarball/wireguard-vxlan
ERROR: Could not install requirement netjsonconfig@ https://github.com/openwisp/netjsonconfig/tarball/wireguard-vxlan from https://github.com/openwisp/netjsonconfig/tarball/wireguard-vxlan (from openwisp-controller==0.9a0) because of HTTP error 404 Client Error: Not Found for url: https://codeload.github.com/openwisp/netjsonconfig/legacy.tar.gz/wireguard-vxlan for URL https://github.com/openwisp/netjsonconfig/tarball/wireguard-vxlan
WARNING: You are using pip version 20.2.4; however, version 22.0.3 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
The command '/bin/sh -c pip install --prefix='/install' --upgrade --force-reinstall ${OPENWISP_CONTROLLER_SOURCE}' returned a non-zero code: 1
make: *** [Makefile:26: base-build] Error 1
```
Same thing can be seen while running `CI build` on my fork :point_down: 
https://github.com/Aryamanz29/docker-openwisp/runs/5432196168?check_suite_focus=true#step:6:2401

Reply:
Yeah, that's not you, the version of openwisp-controller broke for some reason, I will fix it today, just rebase after that! 😄

Reply:
Updated repo.
If it doesn't work on local, please clear your docker cache (you can use `--no-cache` option in make build step, let me know if I should explain in detail)! 😄

Reply:
It works 🙂 
- It seems like `drf-yasg` doesn't support `django~=4.0.0`, It's giving me error :point_down: 
Related to : https://github.com/axnsan12/drf-yasg/issues/761, But I found that we've already updated `CompressStaticFilesStorage` class in https://github.com/openwisp/openwisp-utils/pull/284/files, then don't know why we're getting same error 🤔

Note :  Static minification and cache invalidation works perfectly fine `django3`

```bash
dashboard_1   |   File "/usr/local/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 381, in _stored_name
dashboard_1   |     cache_name = self.clean_name(self.hashed_name(name))
dashboard_1   |   File "/usr/local/lib/python3.10/site-packages/openwisp_utils/storage.py", line 19, in hashed_name
dashboard_1   |     return super().hashed_name(name, content, filename)
dashboard_1   |   File "/usr/local/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 106, in hashed_name
dashboard_1   |     raise ValueError(
dashboard_1   | ValueError: The file 'drf-yasg/redoc/redoc.standalone.js.map' could not be found with <openwisp_utils.storage.CompressStaticFilesStorage object at 0x7fe970ab4f10>.
```

Reply:
I recommend going inside the image and checking the version. (i.e use `docker exec -it docker-compose<container-name>`)

P.S: To find container name, you might wanna use `docker images`

Also, there is a vscode extension called `remote` -> `container` with which you can get inside running docker containers with easy.

I could help you with it by doing it on my end, but that would take a couple of days!

Reply:
Thanks, @atb00ker :smile:   
After going inside the image, I found `openwisp-utils` code is not updated with its current version (It Updated last on 5 Jan 2022), that's why it raises above error on `django4`. We need to point `OPENWISP_UTILS_SOURCE` to it's current release version.

```bash
root@685413f567f9:/usr/local/lib/python3.10/site-packages/openwisp_utils# cat storage.py 
```
```python
import fnmatch

from compress_staticfiles.storage import (
    CompressStaticFilesStorage as BaseCompressStaticFilesStorage,
)
from django.conf import settings


class FileHashedNameMixin:
    default_excluded_patterns = ['leaflet/*/*.png']
    excluded_patterns = default_excluded_patterns + getattr(
        settings, "OPENWISP_STATICFILES_VERSIONED_EXCLUDE", []
    )

    def hashed_name(self, name, content=None, filename=None):
        if not any(
            fnmatch.fnmatch(name, pattern) for pattern in self.excluded_patterns
        ):
            return super().hashed_name(name, content, filename)
        return name


class CompressStaticFilesStorage(
    FileHashedNameMixin,
    BaseCompressStaticFilesStorage,
):
    """
    A static files storage backend for compression that inherits from
    django-compress-staticfiles's CompressStaticFilesStorage class;
    also adds support for excluding file types using
    "OPENWISP_STATICFILES_VERSIONED_EXCLUDE" setting.
    """

    pass
```
Current `storage.py` https://github.com/openwisp/openwisp-utils/blob/master/openwisp_utils/storage.py, Which resolve this issue.

Reply:
That must be because [docker-openwisp-usage branch](https://github.com/openwisp/openwisp-utils/blob/docker-openwisp-usage/openwisp_utils/storage.py) is used by docker-openwisp.

*Why is this branch used?*

Because if I use master things will keep breaking left and right for people who are using these images and want to build images themselves for customization.

*How can we update this branch?*

1. Easy, check the base image in your local base image: `build/openwisp_base/Dockerfile` and find the line installing openwisp-utils.
2. Use `master` branch instead of the `docker-openwisp-usage` branch
3. Do `make develop` to build the images.
4. Test manually and ensure `make develop-runtests` pass. The bottom line is that everything should be working after the update and all the pages should load as expected.
5. If all works, create a PR to update `docker-openwisp-usage` or ask me to update it. If something breaks, we have to fix it first.
6. Clear docker cache and do `make develop` after the branch is updated and you should be gold! 😄 

Btw, great job finding the exact problem, I appreciate it! 😄

Reply:
Thanks, @atb00ker, for your in-depth explanation, Now things clear to me :smile: :+1:

Reply:
Tested with `OPENWISP_UTILS_SOURCE=https://github.com/openwisp/openwisp-utils/tarball/master` 
`**openwisp-utils update  docker-openwisp-usage PR** ` :arrow_right:  https://github.com/openwisp/openwisp-utils/pull/287  @atb00ker 

![Screenshot from 2022-03-10 00-27-14](https://user-images.githubusercontent.com/56113566/157522317-502c4c36-5d5e-4184-b692-e38c04f17f42.png)

`staticfiles minification`

![Screenshot from 2022-03-10 01-13-04](https://user-images.githubusercontent.com/56113566/157522637-02a87aff-a6e2-4947-a14b-e5e09674f7a2.png)

Reply:
I see that version `1.0.0` was released 20 days ago,
you can just do `openwisp-utils~=1.0.0` in the docker image and we should be good! 😄

Reply:
Done :+1: